### ISSUE TITLE
Make moz-rewrite work again with latest Firefox

### ISSUE THREAD

**Issue by baptx** [warren-bank/moz-rewrite#11]
Hi,

I saw that you don't want to update the moz-rewrite Firefox addon anymore after Mozilla announced that they will deprecate the XUL / XPCOM addon APIs:
https://addons.mozilla.org/en-US/firefox/addon/moz-rewrite-json/

I am also not a fan of removing completely the ability to install unsigned addons. But I can understand that average users don't need to install unsigned addons and it is an additional security that can't be bypassed if it is forbidden. At least it is possible to use an unbranded version of Firefox that allows unsigned addons:
https://wiki.mozilla.org/Add-ons/Extension_Signing#Unbranded_Builds

The ModifyHeaders addon still works with latest Firefox version but unfortunately not moz-rewrite which was able to edit both HTTP request and response headers coming from and going to any domains.

See how moz-rewrite was useful for example to block all HTTP Referer request headers except the ones coming from twitter.com that are going to twitter.com (Twitter needs Referer header to work but we can remove it for third-party domains for privacy reasons):
http://pastebin.com/zgPNYmp2

I don't think there is any addon equivalent today, is there?
Don't you have any plans to make your addon work again?
Is switching to WebExtensions or Addon-SDK (Jetpack) APIs too complicated?

Thanks.

**Comment by warren-bank** [comment]
Hi @baptx 

I'm glad you've found this addon useful. On the one-hand, I'd like to maintain some version of this code. On the other hand, I'm unwilling to dedicate the amount of effort required to do a major rewrite every time that Mozilla decides to pivot and makes major breaking changes to its API.

With respect to the current issue, I believe that it traces to a problem identified by another observant user in a [comment on AMO](https://addons.mozilla.org/en-US/firefox/addon/moz-rewrite-js/reviews/827283/). Here's a copy/paste:

``` text
Bug Fix
by Frank Einstein on October 29, 2016

If this addon is not working for you... Try this!

Open "rewrite_http_headers_js-1.01-fx.xpi" with your favorite archive manager like WinRar or 7-Zip.
(Or, instead, you can download the sources files. It's very easy to build.
Sources: https://github.com/warren-bank/moz-rewrite )

Open this file:
    \components\04_modular_addon\Shared_Sandbox.js

Go to line 89 and replace this line:
    s = [toHexString(hash.charCodeAt(i)) for (i in hash)].join("");

With these 3 lines:
    for (var i = 0; i < hash.length; i++){
        s += toHexString(hash.charCodeAt(i));
    }

Save "Shared_Sandbox.js"
Save the modifications in "rewrite_http_headers_js-1.01-fx.xpi"
(Run Build.bat if you downloaded the sources.)

Install "rewrite_http_headers_js-1.01-fx.xpi"
(to install.. open the xpi file with Firefox or drag-n-drop it on the Firefox window.)

That's it! It should be working now. :)

This addon was always working with my previous Firefox installations but recently I installed this on Firefox v49 for Linux and it wasn't working at all. After a quick look at the main console, I discovered this bug. I think the bug comes from Firefox rather than this addon but it's way more easier to modify the addon code than fixing and recompiling Firefox!
```

..to which I responded:

``` text
nice catch..
by Warren Bank (Developer) on November 7, 2016

https://bugzilla.mozilla.org/show_bug.cgi?id=1220564

https://developer.mozilla.org/en-US/Firefox/Releases/46#InterfacesAPIsDOM
    (scroll up one line)

what's funny is.. that line of code was copied from some dark corner of the Firefox documentation; it's definitely not in my coding style.

a quick search on Google shows lots of other Firefox addon projects that use the exact same snippet of code.. so I guess they all broke with the v46 release. nice FF...

update: found it..

per a discussion found here:
    https://groups.google.com/forum/#!topic/provider-for-google-calendar/Rm2F0UtV3xU

ironically, on a topic having nothing to do with this issue..
they provide a complete verbatim copy of the original documentation,
as well as a link to where it came from:
    https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsICryptoHash

it seems they've since updated the code on that page to read:
    var s = Array.from(hash, (c, i) => toHexString(hash.charCodeAt(i))).join("");

in any case, if your change works.. that's great! Nice catch and thank you.

PS: does your suggestion to unpack, modify, and repack the addon work?
I thought that Firefox began to disallow the installation of UNSIGNED addons with the v48 release.
(actually v43, but at this point there was still a preference that could be changed to override the behavior; now there is not).
By "signed", I mean signed by Mozilla via their "addons" website..
which was effectively the death of writing/distributing private-party addons.
    https://wiki.mozilla.org/Add-ons/Extension_Signing
```

I suppose it wouldn't be much effort to commit this one change and sign the resulting addon using the new [web service signing API](https://blog.mozilla.org/addons/2015/11/20/signing-api-now-available/). The signed .xpi could then be attached to a new [release](https://github.com/warren-bank/moz-rewrite/releases), and anybody interested could just download it from there.

As for the question of future development, the new [WebExtensions API](https://developer.mozilla.org/en-US/Add-ons/WebExtensions) does have:
- [webRequest](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/webRequest) to intercept and conditionally modify both request and response HTTP headers
- [storage.StorageArea](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/StorageArea) to [store key/value pairs](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/storage/StorageArea/set) (up to 1MB per value)

but it lacks:
- access to the filesystem
  - read rules from external .js or JSON files
  - save() requests
- ability to spawn another process
  - `wget` saved requests

All request/response rules would then need to be managed using some GUI that reads/writes the data from/to the "storage" API.

Chrome already has an extension, named ["ModHeader"](https://chrome.google.com/webstore/detail/modheader/idgpnmonknjnojddfkpgkljpfnnfcklj?hl=en), that (though I haven't actually tried it..) appears to have similar behavior.  Rather than my rewriting my addon to look and feel more like this other one, it would probably be a lot easier to massage this other extension to work as a WebExtensions addon for Firefox. Firefox has made it clear that they want their API to be a subset of that of Chrome.. with the intention that they discard the Firefox ecosystem of addons in exchange for access to the Chrome ecosystem via an inter-operable/compatible API. I'm [not happy](http://www.zdnet.com/article/mozilla-changes-firefox-apis-developers-unhappy/) about it either.. but it is what [it is](https://blog.mozilla.org/addons/2015/08/21/the-future-of-developing-firefox-add-ons/).

**tl;dr** - I'll make one quick update to get this addon working again in the current release of Firefox. It was announced on 08/21/2015 that XPCOM and XUL would be deprecated within the next 12 to 18 months. According to my math skills, (carry the one..) that could be any day now. As such, I wouldn't expect many more updates in the future.

**Comment by warren-bank** [comment]
PS:
I like your idea for removing the `referer` header from cross-domain requests.
While I'm touching code, I think I'll add the following (not-yet-tested) recipe:

``` javascript
[
    {
        "url" : /^.*$/,
        "headers" : function() {
            var headers = {};

            // remove referer from all cross-domain requests
            if ( (request.headers.unmodified.referer) && (request.headers.unmodified.referer !== request.window_location.host) ){
                headers.referer = null;
            }

            return headers;
        }
    }
]
```

It takes your idea and makes it host agnostic. The header is sent back to its own server, but is removed from requests made (by a webpage) to other hosts.

**Comment by warren-bank** [comment]
oops.. this is more like it :)

``` javascript
[
    {
        "url" : /^.*$/,
        "headers" : function() {
            var headers = {};
            var host_regex_pattern, host_regex;

            // remove referer from all cross-domain requests
            if (request.headers.unmodified.referer){
                host_regex_pattern = '^https?://(?:[^\./]+\.)*?' + request.window_location.host + '(?:/.*)?$';
                host_regex = new RegExp(host_regex_pattern, 'i');

                if (host_regex.test(request.headers.unmodified.referer) == false){
                    headers.referer = null;
                }
            }

            return headers;
        }
    }
]
```

**Comment by warren-bank** [comment]
i admit, i hadn't read the docs since i wrote them..
noticing that there's a [`request.referrer`](https://github.com/warren-bank/moz-rewrite/blob/c2acfb39579faa923ede1c1d51d465da014735e2/components/04_modular_addon/HTTP_Stream.js#L375) value that could be used, here's (yet another) version..

``` javascript
[
    {
        "url" : /^.*$/,
        "headers" : function() {
            var headers = {};

            // remove referer from all cross-domain requests
            if ( (request.headers.unmodified.referer) && (request.referrer) && (request.referrer.host !== request.window_location.host) ){
                headers.referer = null;
            }

            return headers;
        }
    }
]
```

**Comment by Gitoffthelawn** [comment]
I **_completely**_ understand your concerns regarding Mozilla's pivots.

Here is my current understanding:
1. Mozilla is _slowly_ switching to Web Extensions and deprecating previous technologies.  I would not expect them to accomplish this anywhere close to the timetable they laid out.
2. Mozilla's Web Extensions will probably be _mostly_ a **superset** of Chrome's implementation.  Mozilla has a long history of allowing their browsers to be customizable, and Chrome's implementation does not sufficiently allow this.
3. It's unlikely that Mozilla will pivot again anytime in the near future.  With multiple mainstream browsers sharing a common base of Web Extensions API's, the standard is likely to used for quite some time.  The extension ecosystem developed around Web Extensions will help ensure it's longevity.

**Comment by warren-bank** [comment]
[here](https://github.com/warren-bank/moz-rewrite/releases/tag/js%2FCu.evalInSandbox%2Fv1.02) is a new working release.. at least in the sense that it works in the "Developer Edition"

[some notes](https://github.com/warren-bank/moz-rewrite/blob/js/Cu.evalInSandbox/master/.scripts/03.%20sign%20xpi/5.%20result.txt) regarding the "signing API" process..

long story short:
- it's not an automated immediate-turnaround "we'll sign anything you want for you to test or distribute yourself.. just don't expect us to host it until we've given it a closer look-see"
- instead, it's more of an "use our complicated API to save yourself the hassle of logging into our website directly"
- all it did was upload the .xpi into the normal queue for manual inspection and approval.. not what I wanted at all.. argh

anyway, I tested it with the following rule applied to "requests":

``` javascript
[
    {
        "url": /^.*$/,
        "headers": function(){
            var $headers = {};
            var val      = 'abcdefghijklmnopqrstuvwxyz';

            var $md5             = md5(val);
            var $sha1            = sha1(val);
            var $base64          = base64_encode(val);

            $headers['x-md5']    = $md5;
            $headers['x-sha1']   = $sha1;
            $headers['x-base64'] = $base64;

            return $headers;
        }
    }
]

/* ********************************************************
 * url: http://headers.jsontest.com/?mime=1
 *
 * request headers:
 * {
 *   "x-md5"    : "c3fcd3d76192e4007dfb496cca67e13b",
 *   "x-sha1"   : "32d10c7b8cf96570ca04ce37f2a19d84240d3a89",
 *   "x-base64" : "YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXo="
 * }
 * ********************************************************
 */
```

I focused on using the crypto functions since I wanted to confirm that the updated code was working.
It seems to be..

**Comment by mikhoul** [comment]
It's available on AMO !   :smile:

**Comment by warren-bank** [comment]
@mikhoul 

You got my hopes up!

It's status is still pending review.
("Queue Position: 65 of 82")

It's listed under ["complete version history"](https://addons.mozilla.org/en-US/firefox/addon/moz-rewrite-js/versions/), but if you try to ["Add to Firefox"](https://addons.mozilla.org/firefox/downloads/file/537173/rewrite_http_headers_js-1.02.1-signed-fx.xpi) you'll quickly realize that it hasn't yet been signed. FF43+ will "prevent this site from installing an unverified add-on".

notes:
- the .xpi that was uploaded using their "signing API" was version: "1.02"
- I logged into AMO and deleted this upload, since i wanted to actually test the .xpi before uploading a finished copy to the public site
- after some toying around in the "Developer Edition" testing unsigned flavors of the .xpi, I pushed the release to github and resubmit the .xpi to AMO (still version: "1.02")
- AMO complained that this is now an invalid version number because they have some record of it having been uploaded previously (though there's no trace of it in my "Developer Hub")
- I simply changed the version string to "1.02.1-signed" and re-uploaded to AMO
- they accepted it and this is the copy that's pending review; it's identical to the .xpi here on github with that one exception.. that the version string was changed to workaround a bug in their website
- if they ever do approve and sign it, they'll change this version string to "1.02.1-signed.1-signed"
- since they have a habit of appending ".1-signed" several times to their published addons, I figured it was an innocuous trick

additional notes:
- I just really hope that submitting this updated version isn't going to backfire
- in the past, I had another addon that had been approved.. made public on AMO.. had users.. and when I pushed a very small change.. a reviewer (for no apparent reason) not only didn't accept the new version, but also retroactively removed the previously accepted versions.. effectively removing the addons from AMO
- ever since that happened, I've been gun-shy about pushing any updates to AMO. Before signing became mandatory, it was just simpler to post unsigned .xpi files as attachments along with each release on github.
- in any case, if they don't approve it.. I don't really care. I just hope they don't pull the addon from the site altogether.
- in a perfect world, I'd wake up in the morning (after submitting it at a late hour on the night prior) and it really would already be available on AMO! :)

**Comment by warren-bank** [comment]
**update:**

Your add-on, Rewrite HTTP Headers (JS) 1.02.1-signed, has been flagged for Admin Review.

Your add-on is still in our review queue, but it will need to be checked by one of our admin reviewers. This means that your review will probably take longer than usual.

**Comment by Gitoffthelawn** [comment]
That sounds like a PITA.

You might find it useful to post your experiences somewhere where someone at Mozilla will also read them.

Perhaps: https://discourse.mozilla-community.org/c/add-ons

**Comment by norstbox** [comment]
Hi @warren-bank 
Looks like `redirectTo` function breaks FireFox UI, at least in version 50.

Regarding signing process. If you are willing to not use an API, you might use [Submit a New Add-on page](https://addons.mozilla.org/en-US/developers/addon/submit/distribution) on Developer Hub and choice _On your own_ as distribution method:
> Your submission will be _immediately signed for self-distribution_. Updates should be handled by you via an updateURL or external application updates.

**Update**. Seems like I have found the cause of the issues you are running into:
> You add-on passed validation but it could not be automatically signed because your add-on is [listed](https://developer.mozilla.org/en-US/Add-ons/Distribution). Listed add-ons will require a manual review before they can be signed.

Please take a look at [jpm man page](https://developer.mozilla.org/en-US/Add-ons/SDK/Tools/jpm#jpm_sign) and [Signing and distributing at MDN](https://developer.mozilla.org/en-US/Add-ons/Distribution).

**Comment by warren-bank** [comment]
Hi @norstbox 
Nice catch.. that was **completely** my fault.
I just pushed an updated [v1.03](https://github.com/warren-bank/moz-rewrite/releases/tag/js%2FCu.evalInSandbox%2Fv1.03) that fixes this bug (that I introduced with the v1.02 commit/release).

Regarding the signing process. I think the reason the web service API didn't work for me.. at least in the sense that the .xpi wasn't automatically/immediately signed and available to download and distribute is due to the fact that I didn't update the UUID of the addon. Presumably, they can't or won't privately sign a test version of an addon that's already publicly listed. At least, that's the conclusion that I came to.. though it doesn't seem to be documented or discussed anywhere.

**Comment by warren-bank** [comment]
* update [v1.04](https://github.com/warren-bank/moz-rewrite/releases/tag/js%2FCu.evalInSandbox%2Fv1.04),<br>adds a `log()` function that passes the parameters to `console.log()` in the active browser window behind the scenes
* update [v1.05](https://github.com/warren-bank/moz-rewrite/releases/tag/js%2FCu.evalInSandbox%2Fv1.05),<br>finally got around to taking another look at `redirectTo()`.. and making it work the way it was supposed to (..for page resources, as well as document-level pages)
* 1.05 was pushed to AMO for review..

**Comment by norstbox** [comment]
@warren-bank 
Thank you for the prompt update. Works like a charm now. Maybe I missed something, but without `log()` function debugging was difficult.

> Presumably, they can't or won't privately (automatically) sign a test version of an addon that's already publicly listed. At least, that's the conclusion that I came to.. _though it doesn't seem to be documented or discussed anywhere_.

I feel that quote from my [previous](https://github.com/warren-bank/moz-rewrite/issues/11#issuecomment-262862995) comment (from the [reference page for jpm](https://developer.mozilla.org/en-US/Add-ons/SDK/Tools/jpm#jpm_sign)) about the same.

**Comment by warren-bank** [comment]
@norstbox 
Glad it's working well for you (in v50.. unbranded?); I tested in Dev Ed v51.0.a2.

You were right, this behavior is documented..
* as much as I (and other fans of the direction javascript is headed) love node and npm, I never saw much need for jpm.. so I never stumbled across that discussion in its manpage regarding xpi signing (via the web service API)
* I did see the [blog post](https://blog.mozilla.org/addons/2015/11/20/signing-api-now-available/) that announced the availability of this web service API, which had enough info in it to use. This post mentions some API documentation, but the [link](https://olympia.readthedocs.org/en/latest/topics/api/signing.html) is now 404.
* since we're discussing the topic, I googled and found where this API documentation [currently lives](http://addons-server.readthedocs.io/en/latest/topics/api/signing.html).. and, in fact, it does contain the explanation:

```text
If validation passed then your add-on will be submitted for review.
In the case of unlisted add-ons this will happen automatically.
If your add-on is listed then it will be reviewed by a human and that will take a bit longer.
You can check the "automated_signing" property
to see if signing will happen automatically, or after a manual review.
```

Please feel free to continue testing and let me know if you find any major bugs. I did notice that (for whatever reason) changes to the addon's preferences didn't seem to apply until after a restart, which is odd.. always worked in the past, but not a big enough issue that I'm going to stress about it.

Fingers X'ed that this update passes the "human review" and finds its way onto AMO some time soon..

**Comment by norstbox** [comment]
@warren-bank 
_in v50.. unbranded?_
I use "general" (release channel build) v50. Of course, the addon id was changed to`moz-rewrite-js_unlisted@warren-bank.github.com` for signing.

_changes to the addon's preferences didn't seem to apply until after a restart_
Not in my case. I wrote several rules over the past few days. Changes to the UI preferences (like enable/disable, watch interval, path etc.) have an immediate effect. I have just tested in Dev v52.0.a2 without issues. OS: Win 8.1.

**Comment by warren-bank** [comment]
nice!

I was trying to figure out how you got an unsigned addon to work in a stable release build number;
never crossed my mind that anybody would be slick enough to roll their own (unlisted / signed) xpi with an updated UUID.

That's good news about the preference updates applying as they should. That was just an initial observation that I had when installing the first iteration of the v1.02 addon.. to give it a quick test. Didn't bother to confirm that observation later on down the line. I'm really glad that either my observation was wrong, or maybe that first iteration had some issue, or maybe the first rule that I had written didn't parse due to some syntax error.. not really sure.. ultimately, it doesn't really matter.. just glad to know that it fixed itself! Those are always the best kinds of bugs :)

Thanks again for all the helpful feedback!

..this same thanks goes to everybody involved in this discussion;
it's nice to see that others have found this addon useful.
